server/jobs/job_manager.go

```diff
@@ -0,0 +1,143 @@
+package jobs
+
+import (
+	"sync"
+	"time"
+
+	"github.com/lieut-data/mattermost-plugin-api/cluster"
+	"github.com/mattermost/mattermost-plugin-mscalendar/server/mscalendar"
+	"github.com/pkg/errors"
+)
+
+type JobManager struct {
+	registeredJobs sync.Map
+	activeJobs     sync.Map
+	env            mscalendar.Env
+	papi           cluster.JobPluginAPI
+}
+
+type RegisteredJob struct {
+	id                string
+	interval          time.Duration
+	work              func(env mscalendar.Env)
+	isEnabledByConfig func(env mscalendar.Env) bool
+}
+
+type activeJob struct {
+	RegisteredJob
+
+	cancel     chan struct{}
+	cancelled  chan struct{}
+	cancelOnce sync.Once
+}
+
+// NewJobManager creates a JobManager for to let plugin.go coordinate with the scheduled jobs.
+func NewJobManager(papi cluster.JobPluginAPI, env mscalendar.Env) *JobManager {
+	return &JobManager{
+		papi: papi,
+		env:  env,
+	}
+}
+
+// AddJob accepts a RegisteredJob, stores it, and activates it if enabled.
+func (jm *JobManager) AddJob(job RegisteredJob) error {
+	jm.registeredJobs.Store(job.id, job)
+	if job.isEnabledByConfig(jm.env) {
+		err := jm.activateJob(job)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// OnConfigurationChange activates/deactivates jobs based on their current state, and the current plugin config.
+func (jm *JobManager) OnConfigurationChange(env mscalendar.Env) error {
+	jm.env = env
+	jm.registeredJobs.Range(func(k interface{}, v interface{}) bool {
+		job, ok := v.(RegisteredJob)
+		if !ok {
+			jm.env.Logger.Errorf("Expected RegisteredJob, got %T", v)
+			return true
+		}
+		enabled := job.isEnabledByConfig(env)
+		active := jm.isJobActive(job.id)
+
+		// Config is set to enable. Job does not exist, create new job.
+		if enabled && !active {
+			err := jm.activateJob(job)
+			if err != nil {
+				jm.env.Logger.Errorf("Error activating job", "id", job.id, "error", err.Error())
+			}
+		}
+
+		// Config is set to disable. Job exists, kill existing job.
+		if !enabled && active {
+			err := jm.deactivateJob(job)
+			if err != nil {
+				jm.env.Logger.Errorf("Error deactivating job", "id", job.id, "error", err.Error())
+			}
+		}
+
+		return true
+	})
+	return nil
+}
+
+// Close deactivates all active jobs. It is called in the plugin hook OnDeactivate.
+func (jm *JobManager) Close() error {
+	jm.activeJobs.Range(func(k interface{}, v interface{}) bool {
+		job := v.(*activeJob)
```

Comment:
A few lines before you checked whether the conversion was possible, and here you are not. Just wondering why the difference.

Reply:
We are only storing `activeJob`s in the map. I may just get rid of the checks, what do you think?

Reply:
Sounds reasonable for me.